### PR TITLE
[KAT-2014] API to add properties to synthetic in-memory property graphs after construction

### PR DESCRIPTION
**Pull request by yan-zaretskiy**

### CODE REVIEW

File: libgalois/include/katana/TopologyGeneration.h [hunk 1]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+
+public:
+  using ValueType = std::invoke_result_t<F, Input>;
+  using ArrowType = typename arrow::CTypeTraits<ValueType>::ArrowType;
+  using BuilderType = typename arrow::TypeTraits<ArrowType>::BuilderType;
+
+  PropertySetter(std::string name, F resolve_value)
+      : name(std::move(name)), resolve_value(std::move(resolve_value)) {}
+
+  std::shared_ptr<arrow::Field> Field() const noexcept {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: minor: rename to MakeField

File: libgalois/include/katana/TopologyGeneration.h [hunk 2]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+    std::shared_ptr<arrow::DataType> type = std::make_shared<ArrowType>();
+    return arrow::field(name, type);
+  }
+
+  std::shared_ptr<BuilderType> Builder() const noexcept {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: rename to MakeBuilder

File: libgalois/include/katana/TopologyGeneration.h [hunk 3]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+    return std::make_shared<BuilderType>();
+  }
+
+  ValueType operator()(Input id) const { return resolve_value(id); }
+
+private:
+  std::string name;
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: we use silly Google Style Guide which uses `name_` for member variables of classes.

File: libgalois/include/katana/TopologyGeneration.h [hunk 4]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+  F resolve_value;
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: rename F to more informative name. Also `resolve_value` should be `value_generator_func_` or similar to clarify that it's a function object.

File: libgalois/include/katana/TopologyGeneration.h [hunk 5]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+struct is_property_setter : is_property_setter_impl<std::decay_t<T>> {};
+
+template <typename... Ts>
+struct all_property_setters : std::conjunction<is_property_setter<Ts>...> {};
+
+template <bool is_node, typename... Args>
+Result<void>
+AddGraphProperties(katana::PropertyGraph* pg, Args&&... setters) {
+  static_assert(
+      internal::all_property_setters<Args...>::value,
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: nice.

File: libgalois/include/katana/TopologyGeneration.h [hunk 6]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+
+  auto append_table_components = [&](const auto& setter) -> Result<void> {
+    // For schema
+    fields.emplace_back(setter.Field());
+
+    // For property values
+    auto builder = setter.Builder();
+    if constexpr (is_node) {
+      builder->Reserve(pg->num_nodes());
+      for (auto node_it = pg->begin(); node_it != pg->end(); ++node_it) {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: `for (Node n: pg->all_nodes()) {`

File: libgalois/include/katana/TopologyGeneration.h [hunk 7]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+        KATANA_CHECKED(builder->Append(setter(*node_it)));
+      }
+    } else {
+      builder->Reserve(pg->num_edges());
+      auto edges = pg->topology().all_edges();
+      for (auto edge_it = edges.begin(); edge_it != edges.end(); ++edge_it) {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: `for (Edge e: pg->topology().all_edges()) {...`
In general, favor range-based for over iterator based.

File: libgalois/include/katana/TopologyGeneration.h [hunk 8]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+    KATANA_CHECKED(builder->Finish(&array));
+    columns.emplace_back(std::make_shared<arrow::ChunkedArray>(array));
+
+    return katana::ResultSuccess();
+  };
+
+  // Iterate over all the supplied PropertySetters and build an arrow::Table.
+  // We fold over || to terminate early if any call returns an error.
+  Result<void> result = katana::ResultSuccess();
+  ((result = append_table_components(std::forward<Args>(setters)),
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: aren't variadic expressions a pleasure? :)

File: libgalois/include/katana/TopologyGeneration.h [hunk 9]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+  static_assert(
+      std::is_invocable_v<F, Input>,
+      "PropertySetter must be constructed with an invokable type.");
+
+public:
+  using ValueType = std::invoke_result_t<F, Input>;
+  using ArrowType = typename arrow::CTypeTraits<ValueType>::ArrowType;
+  using BuilderType = typename arrow::TypeTraits<ArrowType>::BuilderType;
+
+  PropertySetter(std::string name, F resolve_value)
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: `const F&` for formal parameter.

File: libgalois/include/katana/TopologyGeneration.h [hunk 10]
```diff
@@ -24,9 +31,143 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+  } else {
+    return pg->AddEdgeProperties(table);
+  }
+}
+
+}  // namespace internal
+
+template <typename F>
+KATANA_EXPORT internal::PropertySetter<PropertyGraph::Node, F>
+NodePropertySetter(std::string name, F resolve_value) {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: `const std::string&` and `const F&` all around, then you don't need `std::move()`.

File: libgalois/test/property-graph-in-memory-props.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+using Edge = katana::PropertyGraph::Edge;
+using Node = katana::PropertyGraph::Node;
+using katana::AddEdgeProperties;
+using katana::AddNodeProperties;
+using katana::EdgePropertySetter;
+using katana::NodePropertySetter;
+
+void
+TestNodeProps(std::unique_ptr<katana::PropertyGraph>&& pg) {
```
Comment (amberhassaan) [inline libgalois/test/property-graph-in-memory-props.cpp:13]: use range-based for loops.

**Review by amberhassaan — COMMENTED** [review]
Mostly stylistic changes.

File: libgalois/include/katana/TopologyGeneration.h [hunk 12]
```diff
@@ -24,9 +31,145 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
     size_t num_rows) noexcept;
+
+/***********************************************************/
+/* Functions for adding node and edge properties to graphs */
+/***********************************************************/
+
+namespace internal {
+
+template <typename Input, typename ValueFunc>
+class PropertySetter {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: Add doxygen comments. What are your thoughts about renaming this `PropertyGenerator`?

File: libgalois/include/katana/TopologyGeneration.h [hunk 13]
```diff
@@ -24,9 +31,145 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+  } else {
+    return pg->AddEdgeProperties(table);
+  }
+}
+
+}  // namespace internal
+
+template <typename ValueFunc>
+KATANA_EXPORT internal::PropertySetter<PropertyGraph::Node, ValueFunc>
+NodePropertySetter(const std::string& name, const ValueFunc& resolve_value) {
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: Add doxygen comments and rename `resolve_value`.

File: libgalois/include/katana/TopologyGeneration.h [hunk 14]
```diff
@@ -24,9 +31,145 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+  return {name, resolve_value};
+}
+
+template <typename ValueFunc>
+KATANA_EXPORT internal::PropertySetter<PropertyGraph::Edge, ValueFunc>
```
Comment (amberhassaan) [inline libgalois/include/katana/TopologyGeneration.h]: same for the rest of the functions. Mainly need doxygen comments with a usage example.

File: libgalois/test/property-graph-in-memory-props.cpp [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+  auto avgs = pg->GetEdgeProperty("average").value();
+  auto avgs_array =
+      std::static_pointer_cast<arrow::DoubleArray>(avgs->chunk(0));
+
+  auto names = pg->GetEdgeProperty("edge_name").value();
+  auto names_array =
+      std::static_pointer_cast<arrow::StringArray>(names->chunk(0));
+
+  size_t i = 0;
+  for (Edge e : pg->topology().all_edges()) {
```
Comment (amberhassaan) [inline libgalois/test/property-graph-in-memory-props.cpp]: Might be a good idea to add `all_edges()` to `PropertyGraph`, then we can use `all_nodes` and `all_edges` uniformly.

File: libgalois/test/property-graph-in-memory-props.cpp [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+  auto ages = pg->GetNodeProperty("age").value();
+  auto ages_array = std::static_pointer_cast<arrow::Int32Array>(ages->chunk(0));
+
+  auto names = pg->GetNodeProperty("name").value();
+  auto names_array =
+      std::static_pointer_cast<arrow::StringArray>(names->chunk(0));
+
+  size_t i = 0;
+  for (Node n : *pg) {
```
Comment (amberhassaan) [inline libgalois/test/property-graph-in-memory-props.cpp]: I personally prefer `all_nodes()` for clarity.

**Review by amberhassaan — APPROVED** [review]
PR is ready IMO. Please add doxygen and address other minor comments.

File: libgalois/include/katana/TopologyGeneration.h [hunk 17]
```diff
@@ -24,9 +31,197 @@ KATANA_EXPORT std::unique_ptr<katana::PropertyGraph> MakeSawtooth(
...
+struct is_property_generator_impl<PropertyGenerator<V>> : std::true_type {};
+
+template <typename T>
+struct is_property_generator : is_property_generator_impl<std::decay_t<T>> {};
+
+template <typename... Ts>
+struct all_property_generators
+    : std::conjunction<is_property_generator<Ts>...> {};
+
+template <bool is_node, typename... Args>
```
Comment (l-hoang) [inline libgalois/include/katana/TopologyGeneration.h:95]: Could you add a function that calls into AddGraphProperties with the correct template argument?

e.g., AddNodeProperties would call AddGraphProps<true>, while AddEdgePropertties would call AddGraphProps<false>.

This makes it so the user does not need to be aware of what the template argument is/should be.

**Review by l-hoang — APPROVED** [review]
my comment is a preference more than anything that needs to get done

File: libgalois/include/katana/TopologyGeneration.h (hunk 17 above)
Comment (yan-zaretskiy) [reply libgalois/include/katana/TopologyGeneration.h:95]: Just to clarify - do you mean something different than the `AddNodeProperties` and `AddEdgeProperties` functions defined further down in the file?

Comment (l-hoang) [reply libgalois/include/katana/TopologyGeneration.h:95]: Didn't notice those: yes, exactly; ignore this comment then

**Pull request merged**